### DISCUSSION TITLE
How to remote debug STM32 code with gdb using stlink-tools on linux?

### DISCUSSION THREAD

**Discussion by fghoussen** [stlink-org/stlink#1361] [Q&A (resolved)] [answered]
On linux, using `stlink-tools`, I have a STM32 board that runs code which must be debugged. How to proceed?

The code is built on host (x86 PC) with debug flag using `gcc-arm-none-eabi` for STM32 target (arm):

```
>> sudo apt-get install gcc-arm-none-eabi
>> make VERBOSE=1
/usr/bin/arm-none-eabi-gcc  ... -g -O0 ... -o blink01.c.o -c blink01.c
/usr/bin/arm-none-eabi-gcc ... -g -O0 blink01.c.o -o blink01.elf
```

Then stripped to reduce the size of the binary:

```
>> /usr/bin/arm-none-eabi-objcopy -O binary blink01.elf blink01.bin

```

Then the reduced binary is flashed to the STM32 board using STM tools: 

```
>> sudo apt install stlink-tools
>> st-flash write blink01.bin 0x08000000
...
INFO common.c: Flash written and verified! jolly good!
```

Then running in one terminal gdbserver:

```
>> st-util
st-util
INFO common.c: F303 high density: 64 KiB SRAM, 512 KiB flash in at least 2 KiB pages.
INFO gdb-server.c: Listening at *:4242...
```

And in another terminal, running (note using .elf on host with debug symbols and .bin on target without symbol as it's supposes to work this way according to a doc I read somewhere):

```
>> sudo apt-get install gdb-arm-none-eabi
>> gdb-multiarch
(gdb) set remote exec-file /blink01.bin
(gdb) file blink01.elf
Reading symbols from blink01.elf...
(gdb) target extended-remote localhost:4242
Remote debugging using localhost:4242
0x00000000 in ?? ()
``` 

Symbols seems to screwed `0x00000000 in ??`: no stack available and breakpoint can be set but are never hit when running `gdb`. Seems symbols can not be used: how to fix this?

**EDIT**

On host:
```
>> uname -a
Linux yoda 6.5.0-4-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.5.10-1 (2023-11-03) x86_64 GNU/Linux
>> arm-none-eabi-gcc --version
arm-none-eabi-gcc (15:12.2.rel1-1) 12.2.1 20221205
>> gdb-multiarch --version
GNU gdb (Debian 13.2-1) 13.2
>> st-flash --version
v1.7.0
```

**Comment by Nightwalker-87** [comment, marked as answer]
As far as I can see the steps taken are not wrong. I suggest to try it out along with example code/binaries from the web and to go through the procedure again.
I'm sorry to say that we can not provide support or any specific documentation for external tools apart from our limited coverage on general use with the `stlink` toolset in our `tutorial.md` to be found in `/doc`. Consulting the `gdb` project on the use of the related tool instead may be a good idea. I believe there already is extensive content generally available regarding debugging with `gdb`.
